### PR TITLE
fix: binary files as part are corrupted when stripping carriage returns

### PR DESCRIPTION
**Pull request by pgmillon**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

When using a simple PNG binary as part of a message, the carriage return characters are ripped off, corrupting the binary.
Not ripping the carriage return characters have an impact in the parts splitting that was updated to use preg_match + PREG_OFFSET_CAPTURE as a replacement to strpos.
This was done the support either \r\n or \n with a regex rather than a more complex approach.
Working with a binary file highlighted the line feed that is added while reading each part, this was removed too a/ to avoid binary corruption and b/ a mime with no line feed shouldn't magically have one when reading.
A unit test was added to ensure further support of binaries.

### CODE REVIEW

File: test/MimeTest.php
Comment (Ocramius) [inline test/MimeTest.php:249]: The changes in the pre-existing tests are a potential BC break :thinking:

File: test/MimeTest.php
Comment (Ocramius) [inline test/MimeTest.php:264]: Same as above: changes to pre-existing test probably hiding a BC break :thinking:

**Review by Ocramius — COMMENTED** [review]
Lacking the ability to evaluate this diff, will ask for a review by other maintainers

**Comment by Ocramius** [comment]
@pgmillon it's possible that this, being a bugfix, should go to `2.9.x`

**Comment by pgmillon** [comment]
Hi,
I wasn't sure how you deal with bugfixes, sorry.
If you pull bugfixes from "legacy" branches into the main branch or if you "backport" from main into legacy branches.
I can update this PR for 2.9.x, as you prefer.

**Comment by Ocramius** [comment]
We merge up from oldstable branches to latest branches.

Since the last release is `2.9.1`, I'd expect this to land in `2.9.2`, therefore `2.9.x` target branch.

We don't merge regular bugs into older branches, since they need to be security issues, in order to qualify for that :) 

EDIT: I changed the base branch, but your commit will need to probably be cherry-picked on top of that, and then force-pushed here, so extraneous commits are gone.

File: test/MimeTest.php
Comment (pgmillon) [reply test/MimeTest.php:249]: This is the base64-encoded version of foobar without this line feed that is magically added when reading a message.
source: https://www.base64encode.net/share/bgVby

Comment (Ocramius) [reply test/MimeTest.php:249]: TIL `Cg==` is `\n` in base64 - probably seen it a million times, and never paid attention :)

**Comment by pgmillon** [comment]
Should be fine for 2.9.x now.

File: test/MimeTest.php
Comment (pgmillon) [reply test/MimeTest.php:264]: I understand your concern, but don't have a clear answer. 
Ripping-off carriage returns is harmful to binaries but people might have built solutions of their own to mitigate it and this fix might break them.

Comment (Ocramius) [reply test/MimeTest.php:264]: TBH, if that's how things are, that would be an https://xkcd.com/1172/ :D 

workflow_2x

I just want a second review by somebody else, but if it's really like that, the bugfix is valid, and to be merged with the test changes too :+1:

**Comment by pgmillon** [comment]
Wait, I just found another bug : if the message sender uses \r\n to generate the message rather than just \n, then the splitMime is not working correctly and leaving a lonely \r.

**Comment by pgmillon** [comment]
There we go : in my original fix I assumed sender's EOL was \n but when testing with a server that was using \r\n, I noticed the intend to prevent binaries corruption was not met because of a trailing \r.
I added a sender EOL detection to fix the parts splitting which now works fine with binaries.

File: test/MimeTest.php
Comment (lcobucci) [reply test/MimeTest.php:249]: That depends on what comes before/later the `\n` btw: https://3v4l.org/Jff9P

**Comment by pgmillon** [comment]
Hi,
any chance to get that fix merged ? I'm waiting for it to cut a new release of a library I work on.
Thanks.

File: src/Decode.php
Comment (weierophinney) [inline src/Decode.php:65]: What does `$matches[0][1]` represent? Can you memoize it into a variable with a descriptive name, please?

File: test/PartTest.php
Comment (weierophinney) [inline test/PartTest.php:209]: Give this a better name: `testBinaryPartsAreCreatedAndParsedWithoutLoss()` or similar.

**Review by weierophinney — CHANGES_REQUESTED** [review]
The changes make sense to me. While I have concerns about the test changes, your arguments for making them make sense, and the liklihood of breakage for users is limited to those who are already pre-processing parts on their own — which will primarily be limited to those trying to fix this very issue.

My main feedback is that I think you should make a few changes to make the "what" more clear: what you are capturing, what behavior you are testing, etc.

**Comment by gsteel** [comment]
This library is being abandoned: https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2023-12-04-TSC-Minutes.md

**Pull request closed without merging**
